go/vt/sidecardb/sidecardb.go

```diff
@@ -0,0 +1,373 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecardb
+
+import (
+	"context"
+	"embed"
+	"fmt"
+	"io/fs"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"vitess.io/vitess/go/stats"
+
+	"vitess.io/vitess/go/mysql/fakesqldb"
+
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/schemadiff"
+)
+
+const (
+	CreateVTDatabaseQuery      = "create database if not exists _vt"
+	UseVTDatabaseQuery         = "use _vt"
+	ShowVTDatabasesQuery       = "SHOW DATABASES LIKE '_vt'"
+	SelectCurrentDatabaseQuery = "select database()"
+	ShowCreateTableQuery       = "show create table _vt.%s"
+	GetCurrentTablesQuery      = "show tables from _vt"
+
+	CreateTableRegexp = "CREATE TABLE .* _vt.*"
+	AlterTableRegexp  = "ALTER TABLE _vt.*"
+)
+
+// all tables needed in _vt sidecar have their schema in the vtschema subdirectory
+//
+//go:embed vtschema/*
+var schemaLocation embed.FS
+
+type vtTable struct {
+	module string // which module uses this table
+	path   string // path of the schema relative to this module
+	name   string // table name
+	schema string // create table dml
+}
+
+func (t *vtTable) String() string {
+	return fmt.Sprintf("_vt.%s (%s)", t.name, t.module)
+}
+
+var vtTables []*vtTable
+var ddlCount *stats.Counter
+
+func init() {
+	initSchemaFiles()
+	ddlCount = stats.NewCounter("SidecarDbDDLQueryCount", "Number of create/upgrade queries executed")
+}
+
+func initSchemaFiles() {
+	err := fs.WalkDir(schemaLocation, ".", func(path string, entry fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !entry.IsDir() {
+			module, name := filepath.Split(path)
+			arr := strings.Split(module, "/")
+			if len(arr) >= 1 {
+				module = arr[1]
+			}
+			name = strings.Split(name, ".")[0]
+			schema, err := schemaLocation.ReadFile(path)
+			if err != nil {
+				panic(err)
+			}
+			vtTables = append(vtTables, &vtTable{name: name, module: module, path: path, schema: string(schema)})
```

Comment:
There is a delicate assumption here that the file name equals the table name. Which _should_ be the case. But we can also easily extract the table name from the `DDLStatement` inside the file. At the very least, we should _validate_ here that they match. This can save some unpleasant surprises down the road.

Reply:
Great suggestion: implemented